File: apps/convoy/src/shortcuts.js
// Keyboard shortcuts.
//
// The workspace stores them in the form the Electron build writes —
// `mod+alt+n` — and every client reads that same file, so the stored notation
// is the one honoured here rather than a translation of it.

import { state } from "./state.js";

/// The actions, their defaults, what they do and where the settings page
/// lists them. Mirrors `SHORTCUT_ACTIONS`, `DEFAULTS` and `DESCRIPTIONS` in
/// convoy-core, and follows the macOS app's bindings where they do not
/// collide with the seven originals.
export const SHORTCUTS = [
  ["palette", "mod+k", "Command palette", "General"],
  ["search", "mod+f", "Find in terminal, or find a project", "General"],
  ["settings", "mod+,", "Settings", "General"],
  ["sidebar", "mod+b", "Toggle sidebar", "General"],
  ["home", "mod+shift+h", "Home", "General"],
  ["dashboard", "mod+alt+d", "Agent Dashboard", "General"],
  ["files", "mod+shift+g", "Files & Changes", "General"],
  ["theme", "mod+alt+t", "Cycle theme", "General"],
  ["wake", "mod+alt+k", "Toggle keep awake", "General"],
  ["newSession", "mod+n", "New session", "Sessions"],
  ["next", "ctrl+tab", "Next tab", "Sessions"],
  ["previous", "ctrl+shift+tab", "Previous tab", "Sessions"],
  ["closeTab", "mod+w", "Close tab", "Sessions"],
  ["switcher", "mod+e", "Switch terminal, recent first", "Sessions"],
  ["reopenTab", "mod+shift+t", "Reopen closed tab", "Sessions"],
  ["resume", "mod+shift+r", "Resume session", "Sessions"],
  ["stop", "mod+.", "Stop session", "Sessions"],
  ["sleep", "mod+alt+z", "Sleep session", "Sessions"],
  ["pin", "mod+alt+p", "Pin or unpin session", "Sessions"],
  ["edit", "mod+i", "Edit name and notes", "Sessions"],
  ["review", "mod+alt+r", "Start review", "Sessions"],
  ["feedback", "mod+shift+b", "Send feedback to builder", "Sessions"],
  ["quick", "mod+/", "Quick commands", "Sessions"],
  ["split", "mod+\\", "Toggle two panes", "Panes"],
  ["layout1", "ctrl+shift+1", "One pane", "Panes"],
  ["layout2", "ctrl+shift+2", "Two panes", "Panes"],
  ["layout4", "ctrl+shift+4", "Four panes", "Panes"],
  ["paneNext", "mod+alt+right", "Focus next pane", "Panes"],
  ["panePrevious", "mod+alt+left", "Focus previous pane", "Panes"],
  ["paneClose", "mod+shift+w", "Close focused pane", "Panes"],
  ["openFolder", "mod+o", "Open folder", "Project"],
  ["newTask", "mod+shift+n", "New task", "Project"],
  ["newSpec", "mod+alt+n", "New specification", "Project"],
  ["importIssues", "mod+shift+i", "Import Linear or Jira issues", "Project"],
  ["sessionsTab", "mod+alt+1", "Sessions", "Project"],
  ["reviewsTab", "mod+alt+2", "Reviews", "Project"],
  ["specsTab", "mod+alt+3", "Specs", "Project"],
  ["tasksTab", "mod+alt+4", "Tasks", "Project"],
  ["docsTab", "mod+alt+5", "Docs & specs", "Project"],
  ["activity", "mod+shift+a", "Activity feed", "General"],
  ["nextTab", "mod+shift+]", "Next section", "Project"],
  ["previousTab", "mod+shift+[", "Previous section", "Project"],
  ["reveal", "mod+alt+f", "Show project folder", "Project"],
  ["projectRefresh", "mod+alt+shift+r", "Refresh projects in this folder", "Project"],
  ["copyPath", "mod+alt+c", "Copy folder path", "Project"],
  ["gitRefresh", "mod+alt+g", "Refresh Git status", "Project"],
  ["limits", "mod+shift+l", "AI Limits", "Limits"],
  ["limitsRefresh", "mod+alt+l", "Refresh AI limits", "Limits"],
];

export const SHORTCUT_GROUPS = ["General", "Sessions", "Panes", "Project", "Limits"];

/// What is bound to an action right now: the saved value if there is one and
/// it is well formed, the default otherwise.
export function binding(action) {
  const saved = state.settings.shortcuts?.[action];
  const fallback = SHORTCUTS.find(([name]) => name === action)?.[1] ?? "";
  return saved && valid(saved) ? saved : fallback;
}

/// The shape the core enforces, so nothing this writes is rejected on read.
/// `mod` (⌘ on macOS, Ctrl elsewhere) with optional ctrl/alt/shift, or `ctrl`
/// alone — but never Ctrl with a letter, which belongs to the terminal.
export const valid = (value) =>
  /^(mod\+(ctrl\+)?(alt\+)?(shift\+)?([a-z0-9,./;\[\]\\]|left|right|up|down|tab)|ctrl\+(alt\+)?(shift\+)?([0-9,./;\[\]\\]|left|right|up|down|tab))$/.test(value);

const isMac = () => /mac/i.test(navigator.platform);

/// The key a press names, independent of what the modifiers type: on a Mac,
/// Option turns N into "˜" and Shift turns ] into "}", so `event.key` alone
/// never matched an Alt or Shift binding.
const CODES = {
  Comma: ",",
  Period: ".",
  Slash: "/",
  Semicolon: ";",
  BracketLeft: "[",
  BracketRight: "]",
  Backslash: "\\",
  ArrowLeft: "left",
  ArrowRight: "right",
  ArrowUp: "up",
  ArrowDown: "down",
  Tab: "tab",
};

export function keyOf(event) {
  const code = event.code ?? "";
  if (/^Key[A-Z]$/.test(code)) return code.slice(3).toLowerCase();
  if (/^Digit[0-9]$/.test(code)) return code.slice(5);
  if (/^Numpad[0-9]$/.test(code)) return code.slice(6);
  if (CODES[code]) return CODES[code];
  return (event.key ?? "").toLowerCase();
}

/// Whether a key press is the one an action is bound to. `mod` is Command
/// on macOS and Control elsewhere; `ctrl` is Control everywhere. Every
/// modifier must match exactly, so ⌘⇧N never fires ⌘N.
export function matches(event, stored) {
  if (!stored) return false;
  const parts = stored.split("+");
  const key = parts.pop();
  const wants = new Set(parts);
  if (!wants.has("mod") && !wants.has("ctrl")) return false;
  const command = isMac() ? event.metaKey : false;
  const control = event.ctrlKey;
  const wantsCommand = isMac() && wants.has("mod");
  const wantsControl = wants.has("ctrl") || (!isMac() && wants.has("mod"));
  return (
    command === wantsCommand &&
    control === wantsControl &&
    !!event.altKey === wants.has("alt") &&
    !!event.shiftKey === wants.has("shift") &&
    keyOf(event) === key
  );
}

/// A key press written in the stored form, or nothing when it is not a
/// binding — a bare key, a modifier on its own, or Ctrl with a letter.
export function capture(event) {
  if (["Control", "Meta", "Shift", "Alt"].includes(event.key)) return null;
  const mac = isMac();
  const mod = mac ? event.metaKey : event.ctrlKey;
  const ctrl = mac && event.ctrlKey;
  if (!mod && !event.ctrlKey) return null;
  const prefix = mod ? `mod+${ctrl ? "ctrl+" : ""}` : "ctrl+";
  const stored = `${prefix}${event.altKey ? "alt+" : ""}${event.shiftKey ? "shift+" : ""}${keyOf(event)}`;
  return valid(stored) ? stored : null;
}

/// How a binding is shown. The stored form is for the file, not for reading.
export function label(stored) {
  if (!stored) return "unset";
  const mac = isMac();
  const names = mac
    ? { mod: "⌘", ctrl: "⌃", alt: "⌥", shift: "⇧" }
    : { mod: "Ctrl", ctrl: "Ctrl", alt: "Alt", shift: "Shift" };
  const keys = { left: "←", right: "→", up: "↑", down: "↓", tab: "Tab" };
  const parts = stored.split("+");
  const key = parts.pop();
  // macOS lists modifiers ⌃⌥⇧⌘; elsewhere they read left to right with a +.
  const order = mac ? ["ctrl", "alt", "shift", "mod"] : ["mod", "ctrl", "alt", "shift"];
  const modifiers = [...new Set(order.filter((name) => parts.includes(name)).map((name) => names[name]))];
  const shown = keys[key] ?? key.toUpperCase();
  return mac ? `${modifiers.join("")}${shown}` : [...modifiers, shown].join("+");
}

File: apps/linux/crates/convoy-core/src/shortcuts.rs
//! Keyboard shortcuts.
//!
//! The workspace stores them in one portable form — `mod+shift+p` — because
//! every build reads the same file. GTK wants `<Primary><Shift>p`, so the two
//! notations are converted here rather than in the UI, and the stored form
//! never changes.

use crate::model::SHORTCUT_ACTIONS;
use crate::patterns::SHORTCUT;
use std::collections::BTreeMap;

/// What each action does, for the shortcut editor and the command palette.
pub const DESCRIPTIONS: &[(&str, &str)] = &[
    ("palette", "Command palette"),
    ("newSession", "New session"),
    ("files", "Files & Changes"),
    ("next", "Next tab"),
    ("previous", "Previous tab"),
    ("settings", "Settings"),
    ("search", "Search projects and sessions"),
    ("resume", "Resume session"),
    ("stop", "Stop session"),
    ("sleep", "Sleep session"),
    ("pin", "Pin or unpin session"),
    ("edit", "Edit name and notes"),
    ("review", "Start review"),
    ("feedback", "Send feedback to builder"),
    ("quick", "Quick commands"),
    ("split", "Toggle two panes"),
    ("openFolder", "Open folder"),
    ("newTask", "New task"),
    ("newSpec", "New specification"),
    ("importIssues", "Import Linear or Jira issues"),
    ("sessionsTab", "Sessions"),
    ("reviewsTab", "Reviews"),
    ("specsTab", "Specs"),
    ("tasksTab", "Tasks"),
    ("docsTab", "Docs & specs"),
    ("closeTab", "Close tab"),
    ("sidebar", "Toggle sidebar"),
    ("layout1", "One pane"),
    ("layout2", "Two panes"),
    ("layout4", "Four panes"),
    ("paneNext", "Focus next pane"),
    ("panePrevious", "Focus previous pane"),
    ("paneClose", "Close focused pane"),
    ("activity", "Activity feed"),
    ("nextTab", "Next section"),
    ("previousTab", "Previous section"),
    ("reveal", "Show project folder"),
    ("copyPath", "Copy folder path"),
    ("gitRefresh", "Refresh Git status"),
    ("theme", "Cycle theme"),
    ("wake", "Toggle keep awake"),
    ("limits", "AI Limits"),
    ("limitsRefresh", "Refresh AI limits"),
];

/// Defaults used when the workspace has no shortcut for an action: the
/// macOS app's bindings. Next and previous session moved to ⌃Tab / ⌃⇧Tab
/// as there, which freed ⌘⌥P for pin and ⌘⌥N for a new specification.
pub const DEFAULTS: &[(&str, &str)] = &[
    ("palette", "mod+k"),
    ("newSession", "mod+n"),
    ("files", "mod+shift+g"),
    ("next", "ctrl+tab"),
    ("previous", "ctrl+shift+tab"),
    ("settings", "mod+,"),
    ("search", "mod+f"),
    ("resume", "mod+shift+r"),
    ("stop", "mod+."),
    ("sleep", "mod+alt+z"),
    ("pin", "mod+alt+p"),
    ("edit", "mod+i"),
    ("review", "mod+alt+r"),
    ("feedback", "mod+shift+b"),
    ("quick", "mod+/"),
    ("split", "mod+\\"),
    ("openFolder", "mod+o"),
    ("newTask", "mod+shift+n"),
    ("newSpec", "mod+alt+n"),
    ("importIssues", "mod+shift+i"),
    ("sessionsTab", "mod+alt+1"),
    ("reviewsTab", "mod+alt+2"),
    ("specsTab", "mod+alt+3"),
    ("tasksTab", "mod+alt+4"),
    ("docsTab", "mod+alt+5"),
    ("closeTab", "mod+w"),
    ("sidebar", "mod+b"),
    ("layout1", "ctrl+shift+1"),
    ("layout2", "ctrl+shift+2"),
    ("layout4", "ctrl+shift+4"),
    ("paneNext", "mod+alt+right"),
    ("panePrevious", "mod+alt+left"),
    ("paneClose", "mod+shift+w"),
    ("activity", "mod+shift+a"),
    ("nextTab", "mod+shift+]"),
    ("previousTab", "mod+shift+["),
    ("reveal", "mod+alt+f"),
    ("copyPath", "mod+alt+c"),
    ("gitRefresh", "mod+alt+g"),
    ("theme", "mod+alt+t"),
    ("wake", "mod+alt+k"),
    ("limits", "mod+shift+l"),
    ("limitsRefresh", "mod+alt+l"),
];

/// `mod+shift+p` → `<Primary><Shift>p`. Returns nothing for anything the
/// stored format does not allow, so a hand-edited file cannot bind nonsense.
pub fn to_accelerator(value: &str) -> Option<String> {
    if !SHORTCUT.is_match(value) {
        return None;
    }
    // The pattern guarantees modifiers first, in order, then one key.
    let parts: Vec<&str> = value.split('+').collect();
    let (key, modifiers) = parts.split_last()?;
    let modifiers: String = modifiers
        .iter()
        .map(|modifier| match *modifier {
            "mod" => "<Primary>",
            "ctrl" => "<Control>",
            "alt" => "<Alt>",
            _ => "<Shift>",
        })
        .collect();
    Some(format!("{modifiers}{}", key_name(key)))
}

/// `<Primary><Shift>p` → `mod+shift+p`, for the editor's key capture.
pub fn from_accelerator(value: &str) -> Option<String> {
    let mut rest = value;
    let mut modifiers = Vec::new();
    let primary = ["<Primary>", "<Control>", "<Ctrl>", "<Meta>"];
    let mut has_primary = false;
    while let Some(start) = rest.strip_prefix('<') {
        let end = start.find('>')? + 1;
        let token = &rest[..end + 1];
        rest = &rest[end + 1..];
        if primary.contains(&token) {
            has_primary = true;
        } else if token.eq_ignore_ascii_case("<alt>") {
            modifiers.push("alt");
        } else if token.eq_ignore_ascii_case("<shift>") {
            modifiers.push("shift");
        } else {
            return None;
        }
    }
    if !has_primary || rest.is_empty() {
        return None;
    }
    let key = stored_key(rest);
    modifiers.sort_by_key(|name| if *name == "alt" { 0 } else { 1 });
    let result = format!(
        "mod+{}{key}",
        modifiers
            .iter()
            .map(|name| format!("{name}+"))
            .collect::<String>()
    );
    SHORTCUT.is_match(&result).then_some(result)
}

/// GTK names punctuation keys; the stored format spells them out.
const KEY_NAMES: [(&str, &str); 12] = [
    ("tab", "Tab"),
    (",", "comma"),
    (".", "period"),
    ("/", "slash"),
    (";", "semicolon"),
    ("[", "bracketleft"),
    ("]", "bracketright"),
    ("\\", "backslash"),
    ("left", "Left"),
    ("right", "Right"),
    ("up", "Up"),
    ("down", "Down"),
];

fn key_name(key: &str) -> String {
    KEY_NAMES
        .iter()
        .find(|(stored, _)| *stored == key)
        .map(|(_, name)| name.to_string())
        .unwrap_or_else(|| key.to_string())
}

fn stored_key(key: &str) -> String {
    let lower = key.to_lowercase();
    KEY_NAMES
        .iter()
        .find(|(_, name)| name.to_lowercase() == lower)
        .map(|(stored, _)| stored.to_string())
        .unwrap_or(lower)
}

/// The accelerator for every action, saved values first and defaults behind.
pub fn resolve(saved: &BTreeMap<String, String>) -> Vec<(&'static str, String, String)> {
    SHORTCUT_ACTIONS
        .iter()
        .map(|action| {
            let stored = saved
                .get(*action)
                .cloned()
                .or_else(|| {
                    DEFAULTS
                        .iter()
                        .find(|(name, _)| name == action)
                        .map(|(_, value)| value.to_string())
                })
                .unwrap_or_default();
            let accelerator = to_accelerator(&stored).unwrap_or_default();
            (*action, stored, accelerator)
        })
        .collect()
}

pub fn description(action: &str) -> &'static str {
    DESCRIPTIONS
        .iter()
        .find(|(name, _)| *name == action)
        .map(|(_, text)| *text)
        .unwrap_or("Action")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn the_two_notations_round_trip() {
        for stored in [
            "mod+k",
            "mod+shift+p",
            "mod+alt+n",
            "mod+alt+shift+k",
            "mod+,",
        ] {
            let accelerator = to_accelerator(stored).expect(stored);
            assert_eq!(
                from_accelerator(&accelerator).as_deref(),
                Some(stored),
                "{stored} became {accelerator}"
            );
        }
        assert_eq!(to_accelerator("mod+shift+p").unwrap(), "<Primary><Shift>p");
        assert_eq!(to_accelerator("mod+,").unwrap(), "<Primary>comma");
    }

    #[test]
    fn anything_the_stored_format_forbids_is_refused() {
        // The order is fixed, the modifier set is fixed, and Primary is required.
        assert!(to_accelerator("mod+shift+alt+p").is_none());
        assert!(to_accelerator("ctrl+p").is_none());
        assert!(to_accelerator("mod+P").is_none());
        assert!(to_accelerator("").is_none());
        assert!(from_accelerator("<Shift>p").is_none());
        assert!(from_accelerator("p").is_none());
        assert!(from_accelerator("<Primary>").is_none());
    }

    #[test]
    fn every_action_resolves_to_something_usable() {
        let resolved = resolve(&BTreeMap::new());
        assert_eq!(resolved.len(), SHORTCUT_ACTIONS.len());
        assert!(resolved
            .iter()
            .all(|(_, _, accelerator)| !accelerator.is_empty()));

        let mut saved = BTreeMap::new();
        saved.insert("palette".to_string(), "mod+shift+k".to_string());
        let resolved = resolve(&saved);
        let palette = resolved
            .iter()
            .find(|(name, ..)| *name == "palette")
            .unwrap();
        assert_eq!(palette.2, "<Primary><Shift>k", "a saved value wins");
    }
}
